File: sample-conference-videochat/sample-conference-videochat/Components/Settings/SessionSettingsViewController/SessionSettingsViewController.m
//
//  SessionSettingsViewController.m
//  sample-conference-videochat
//
//  Created by Injoit on 25.06.15.
//  Copyright (c) 2015 QuickBlox Team. All rights reserved.
//

#import "SessionSettingsViewController.h"
#import "Settings.h"

@interface SessionSettingsViewController()

@property (weak, nonatomic) IBOutlet UILabel *versionLabel;
@property(strong, nonatomic) Settings *settings;

@end

typedef NS_ENUM(NSUInteger, SessionConfigureItem) {
    
    SessionConfigureItemVideo,
    SessionConfigureItemAuido,
};

@implementation SessionSettingsViewController

- (void)viewDidLoad {
    [super viewDidLoad];

    NSString *appVersion = NSBundle.mainBundle.infoDictionary[@"CFBundleShortVersionString"];
    NSString *appBuild = NSBundle.mainBundle.infoDictionary[@"CFBundleVersion"];
    NSString *version = [NSString stringWithFormat:
                         @"Sample version %@ build %@.\n"
                         "QuickBlox WebRTC SDK: %@ Revision %@",
                         appVersion, appBuild,
                         QuickbloxWebRTCFrameworkVersion, QuickbloxWebRTCRevision];
    
    self.versionLabel.text = version;
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    [self.tableView reloadData];
}

#pragma mark - Actions

- (IBAction)pressDoneBtn:(id)sender {
    //APPLY SETTINGS
    Settings *settings = [[Settings alloc] init];
    [settings applyConfig];
    [settings saveToDisk];
    [self dismissViewControllerAnimated:YES completion:nil];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    
    UITableViewCell *cell = [super tableView:tableView cellForRowAtIndexPath:indexPath];
    cell.detailTextLabel.text = [self detailTextForRowAtIndexPaht:indexPath];
    
#if (TARGET_IPHONE_SIMULATOR)
    // make video setting cell unavailable for sims
    if (indexPath.row == SessionConfigureItemVideo
        && indexPath.section == 0) {
        cell.userInteractionEnabled = NO;
    }
#endif
    
    return cell;
}

#pragma mark - UITableViewDelegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    
    [tableView deselectRowAtIndexPath:indexPath animated:NO];
    
    UITableViewCell *cell = [tableView cellForRowAtIndexPath:indexPath];
    if ([cell.reuseIdentifier isEqualToString:@"LogoutCell"]) {
        
        UIAlertController *alertController =
        [UIAlertController alertControllerWithTitle:nil
                                            message:NSLocalizedString(@"Logout ?", nil)
                                     preferredStyle:UIAlertControllerStyleAlert];
        
        [alertController addAction:[UIAlertAction actionWithTitle:NSLocalizedString(@"Yes", nil)
                                                            style:UIAlertActionStyleDefault
                                                          handler:^(UIAlertAction * _Nonnull action)
                                    {
                                        [self.delegate settingsViewController:self didPressLogout:cell];
                                    }]];
        
        [alertController addAction:[UIAlertAction actionWithTitle:NSLocalizedString(@"NO", nil)
                                                            style:UIAlertActionStyleDefault
                                                          handler:nil]];
        
        [self presentViewController:alertController animated:YES completion:nil];
    }
}

- (NSString *)detailTextForRowAtIndexPaht:(NSIndexPath *)indexPath {
    
    
    if (indexPath.row == SessionConfigureItemVideo) {
        
#if !(TARGET_IPHONE_SIMULATOR)
        return [NSString stringWithFormat:@"%tux%tu", self.settings.videoFormat.width, self.settings.videoFormat.height];
#else
        return @"unavailable";
#endif
        
    }
    else if (indexPath.row == SessionConfigureItemAuido) {
        
        return @"";
    }
    
    return @"Unknown";
}

@end

File: sample-conference-videochat-swift/sample-conference-videochat-swift/sample-conference-videochat-swift/Components/Settings/SessionSettingsViewController/SessionSettingsViewController.swift
//
//  SessionSettingsViewController.swift
//  sample-conference-videochat-swift
//
//  Created by Vladimir Nybozhinsky on 04.10.2018.
//  Copyright © 2018 QuickBlox. All rights reserved.
//

import UIKit
import QuickbloxWebRTC

protocol SettingsViewControllerDelegate: class {
    func settingsViewController(_ vc: SessionSettingsViewController?, didPressLogout sender: Any?)
}

enum SessionConfigureItem : Int {
    case video
    case auido
}


class SessionSettingsViewController: UITableViewController {
    
    @IBOutlet private weak var versionLabel: UILabel!
    private var settings: Settings?
    weak var delegate: SettingsViewControllerDelegate?

    override func viewDidLoad() {
        super.viewDidLoad()
        
        settings = Settings.instance
        
        let appVersion = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String
        let appBuild = Bundle.main.infoDictionary?["CFBundleVersion"] as? String
        let version = """
        Sample version \(appVersion ?? "") build \(appBuild ?? "").\n\
        QuickBlox WebRTC SDK: \(QuickbloxWebRTCFrameworkVersion) Revision \(QuickbloxWebRTCRevision)
        """
        
        versionLabel.text = version
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        tableView.reloadData()
    }
    
    // MARK: - Actions
    @IBAction func pressDoneBtn(_ sender: Any) {
        debugPrint("settings \(String(describing: settings))")
        settings?.saveToDisk()
        settings?.applyConfig()
        dismiss(animated: true)
    }
    
    // MARK: - Table view data source
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        
        let cell: UITableViewCell = super.tableView(tableView, cellForRowAt: indexPath)
        cell.detailTextLabel?.text = detailTextForRow(atIndexPaht: indexPath)
        
        #if targetEnvironment(simulator)
        // Simulator
        if indexPath.row == SessionConfigureItem.video.rawValue && indexPath.section == 0 {
            cell.isUserInteractionEnabled = false
        }
        #endif
        
        return cell
    }

    // MARK: - UITableViewDelegate
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        
        tableView.deselectRow(at: indexPath, animated: false)
        
        let cell: UITableViewCell? = tableView.cellForRow(at: indexPath)
        if (cell?.reuseIdentifier == "LogoutCell") {
            
            let alertController = UIAlertController(title: nil, message: NSLocalizedString("Logout ?", comment: ""), preferredStyle: .alert)
            
            alertController.addAction(UIAlertAction(title: NSLocalizedString("Yes", comment: ""), style: .default, handler: { action in
                self.delegate?.settingsViewController(self, didPressLogout: cell)
            }))
            
            alertController.addAction(UIAlertAction(title: NSLocalizedString("NO", comment: ""), style: .default, handler: nil))
            
            present(alertController, animated: true)
        }
    }
    
    func detailTextForRow(atIndexPaht indexPath: IndexPath?) -> String? {
        
        if indexPath?.row == SessionConfigureItem.video.rawValue {
            
            #if targetEnvironment(simulator)
            // Simulator
            return "unavailable"
            #else
            // Device
            return String(format: "%tux%tu", settings?.videoFormat?.width ?? 640, settings?.videoFormat?.height ?? 480)
            #endif

        } else if indexPath?.row == SessionConfigureItem.auido.rawValue {
            
            return ""
        }
        
        return "Unknown"
    }
}
